### ISSUE TITLE
Active Storage direct uploads

### ISSUE THREAD

**Issue by gahia** [hotwired/turbo#243]
Working with Active Storage direct uploads when the form is contained by a turbo frame seems not to working as expected.

```
<turbo-frame ...>
  <form ...>
    <input type="file" data-direct-upload-url="...">
    <input type="submit" ...>
  </form>
</turbo-frame>
```

Turbo intercepts the form submission and doesn't allow the direct upload run before do the submit request.

![imagen](https://user-images.githubusercontent.com/8942202/113497320-1ac33600-952d-11eb-899b-568fef8d2286.png)

But, when a form like this is NOT contained by a turbo-frame tag, it works as expected: the file is uploaded and then the form submitted by turbo. 

![imagen](https://user-images.githubusercontent.com/8942202/113497323-2878bb80-952d-11eb-907c-3a8ed706bdec.png)

Is there a workaround or possible solution for this?

_Using turbo-rails (0.5.9) by the assets pipeline_

**Comment by javan** [comment]
[`<turbo-frame>`’s form interceptor](https://github.com/hotwired/turbo/blob/88e42cd0e8106fde4166618975683970d1ce7844/src/core/frames/form_interceptor.ts#L23-L33) needs to respect already canceled (`event.defaultPrevented`) `submit` events like the top-level submit observer: https://github.com/hotwired/turbo/blob/88e42cd0e8106fde4166618975683970d1ce7844/src/observers/form_submit_observer.ts#L33-L34

**Comment by seanpdoyle** [comment]
Does this behavior occur on https://github.com/hotwired/turbo/pull/232? It reused FormSubmitObserver instead of FormInterceptor.

**Comment by javan** [comment]
Almost! Active Storage [listens for `submit` on `document`](https://github.com/rails/rails/blob/58ce6907112af45467b9e003763044e634307b87/activestorage/app/javascript/activestorage/ujs.js#L12) so it cancels the event “too late”. Could fix that in Rails with:
```diff
--- a/activestorage/app/javascript/activestorage/ujs.js
+++ b/activestorage/app/javascript/activestorage/ujs.js
@@ -9,7 +9,7 @@ export function start() {
   if (!started) {
     started = true
     document.addEventListener("click", didClick, true)
-    document.addEventListener("submit", didSubmitForm)
+    document.addEventListener("submit", didSubmitForm, true)
     document.addEventListener("ajax:before", didSubmitRemoteElement)
   }
 }
```
Regardless, #232 seems like a good change to make.

**Comment by seanpdoyle** [comment]
@javan I've opened https://github.com/rails/rails/pull/41960 and listed you as a co-author. Is the commit message an accurate description of the issue?

**Comment by seanpdoyle** [comment]
https://github.com/rails/rails/pull/41960#issuecomment-948857795 has been merged and "released" as part of `7.0.0.alpha1` and `7.0.0.alpha2`.

Does that resolve this issue?

**Comment by kylekeesling** [comment]
I just tested this and while it seems to work it also appears that the way to setup ActiveStorage in javascript has changed as I'm getting the following error:

`Uncaught TypeError: ActiveStorage.start is not a function`

Changing my setup to the following seems to do the trick though:

```
//Old
import * as ActiveStorage from "@rails/activestorage"
ActiveStorage.start()

//New
import "@rails/activestorage"
```

The Rails docs/guides don't seem to reflect this change yet, so maybe it's worth me submitting a PR to update the docs?

**Comment by CharlieIGG** [comment]
> [rails/rails#41960 (comment)](https://github.com/rails/rails/pull/41960#issuecomment-948857795) has been merged and "released" as part of `7.0.0.alpha1` and `7.0.0.alpha2`.
> 
> Does that resolve this issue?

I upgraded my package.json to run `"@rails/activestorage": "7.0.0alpha2"`
The behavior has changed in that the Direct Upload events do fire now, but Turbo is still capturing and submitting the form before the Direct Upload has time to finish. Removing the Turbo Frame does give me the expected behavior for Direct Uploads  using the Alpha version on top of my Rails 6 app.

**Comment by seanpdoyle** [comment]
Thank you for following up on this @CharlieIGG. I've opened https://github.com/hotwired/turbo/pull/459 to make the changes mentioned by https://github.com/hotwired/turbo/issues/243#issuecomment-818730045.
